cdap-common/src/main/java/co/cask/cdap/common/io/CaseInsensitiveEnumTypeAdapterFactory.java

```diff
@@ -35,22 +35,37 @@
  * instead of deserializing lowercase only.
  */
 public class CaseInsensitiveEnumTypeAdapterFactory implements TypeAdapterFactory {
+
+  private final boolean useUpperCase;
+
+  public CaseInsensitiveEnumTypeAdapterFactory(boolean useUpperCase) {
+    this.useUpperCase = useUpperCase;
+  }
+
   public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
     @SuppressWarnings("unchecked")
     Class<T> rawType = (Class<T>) type.getRawType();
     if (!rawType.isEnum()) {
       return null;
     }
 
-    final Map<String, T> lowercaseToConstant = new HashMap<>();
+    final Map<String, T> stringToConstants = new HashMap<>();
     for (T constant : rawType.getEnumConstants()) {
-      lowercaseToConstant.put(toLowercase(constant), constant);
+      if (useUpperCase) {
```

Comment:
it would be a little better if you have a `Function<Object, String> normalizeFunction` that you set in the constructor to uppercase or lowercase. That way you don't need this extra comparison, and can always call `normalizeFunction.apply(constant)`